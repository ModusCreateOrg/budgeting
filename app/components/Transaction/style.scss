@import 'theme/variables';

.title {
  margin-bottom: 0;
}

.negative {
  color: $red;
}

.positive {
  color: $green;
}
