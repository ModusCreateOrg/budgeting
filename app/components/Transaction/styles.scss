@import 'theme/variables';

.transaction {
  position: relative;
}

.backButton {
  align-items: center;
  background-color: transparent;
  border: 0;
  color: $blue;
  cursor: pointer;
  display: flex;
  font-size: 1.5rem;
  position: absolute;
  text-decoration: none;

  & > span {
    font-size: 3rem;
    margin-right: 1rem;
  }
}