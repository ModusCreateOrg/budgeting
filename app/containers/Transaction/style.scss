@import 'theme/variables';

.transaction {
  display: flex;
  flex-direction: column;
  button {
    box-shadow: 1px 1px 2px 1px black;
    padding: 10px;
    text-align: center;
    margin-bottom: 10px;
  }
}
