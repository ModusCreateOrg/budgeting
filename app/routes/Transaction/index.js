// @flow
import React from 'react';
import Chunk from 'components/Chunk';

const loadTransactionContainer = () => import('components/Transaction');

const Transaction = props => <Chunk load={loadTransactionContainer} {...props} />;

export default Transaction;
