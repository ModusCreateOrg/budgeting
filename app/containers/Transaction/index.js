// @flow
import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { injectAsyncReducers } from 'store';
import type { Transaction as TransactionType } from 'modules/transactions';
import transactionReducer from 'modules/transactions';
import { getRemainingFlowAsTransaction, getTransaction } from 'selectors/transactions';
import PieChart from 'components/PieChart';

import style from './style.scss';

injectAsyncReducers({
  transactions: transactionReducer,
});

type ItemProps = {
  transaction: TransactionType,
  remainingFlow: TransactionType,
  data: TransactionType[],
  onBackClick: () => void,
};

class Transaction extends React.Component<ItemProps> {
  render() {
    const { transaction, remainingFlow, onBackClick } = this.props;

    const isInflow = transaction.value > 0;
    const absTransactionValue = Math.abs(transaction.value);
    const percentage = absTransactionValue * 100 / (remainingFlow.value + absTransactionValue);
    const twoDecimalsPercentage = Math.round(percentage * 100) / 100;

    const percentageClass = isInflow ? style.inflowPercentage : style.outflowPercentage;

    const data = [
      {
        ...transaction,
        value: absTransactionValue,
      },
      remainingFlow,
    ];

    return (
      <div className={style.transaction}>
        <button onClick={onBackClick}>Back</button>
        <h2>{transaction.description}</h2>
        <h4 className={percentageClass}>{`${isInflow ? '+' : '-'}${twoDecimalsPercentage}%`}</h4>
        <PieChart data={data} dataLabel="description" dataKey="description" dataValue="value" />
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  transaction: getTransaction(state, ownProps),
  remainingFlow: getRemainingFlowAsTransaction(state, ownProps),
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  onBackClick: ownProps.history.goBack,
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Transaction));
