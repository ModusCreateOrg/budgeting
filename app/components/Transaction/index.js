// @flow
import * as React from 'react';
import { Link } from 'react-router-dom';
import TransactionData from 'containers/TransactionData';
import styles from './styles.scss';

type TransactionProps = {
  // history: {
  //   length: number, // The number of entries in the history stack
  //   action: string, // The current action (PUSH, REPLACE, or POP)
  //   location: {
  //     // The current location. May have the following properties:
  //     pathname: string, // The path of the URL
  //     search: string, // The URL query string
  //     hash: string, // The URL hash fragment
  //     state: Object, // location-specific state that was provided to e.g. push(path, state) when this location was pushed onto the stack. Only available in browser and memory history.
  //   },
  //   push: (path: string, state?: Object) => void, // Pushes a new entry onto the history stack
  //   replace: (path: string, state?: Object) => void, // Replaces the current entry on the history stack
  //   go: (position: number) => void, // Moves the pointer in the history stack by n entries
  //   goBack: () => void, // Equivalent to go(-1)
  //   goForward: () => void, // Equivalent to go(1)
  //   block: (prompt: any) => void, // Prevents navigation (see the history docs)
  // },
  // location: {
  //   pathname: string, // The path of the URL
  //   search: string, // The URL query string
  //   hash: string, // The URL hash fragment
  //   state: Object, // location-specific state that was provided to e.g. push(path, state) when this location was pushed onto the stack. Only available in browser and memory history.
  // },
  match: {
    params: Object, // Key/value pairs parsed from the URL corresponding to the dynamic segments of the path
    isExact: boolean, // true if the entire URL was matched (no trailing characters)
    path: string, // The path pattern used to match. Useful for building nested <Route>s
    url: string, // The matched portion of the URL. Useful for building nested <Link>s
  },
};

const Transaction = (props: TransactionProps) => {
  const { id } = props.match.params;

  return (
    <div className={styles.transaction}>
      <Link className={styles.backButton} to="/budget">
        <span>{'<'}</span>
        Back
      </Link>
      <TransactionData id={Number(id)} />
    </div>
  );
};

export default Transaction;
