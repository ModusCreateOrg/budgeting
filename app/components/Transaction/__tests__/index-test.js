import React from 'react';
import renderer from 'react-test-renderer';
import Transaction from 'components/Transaction';

jest.mock('containers/TransactionData', () => 'div');
jest.mock('react-router-dom');

it('renders correctly', () => {
  const tree = renderer.create(<Transaction match={{ params: { id: '4' } }} />).toJSON();
  expect(tree).toMatchSnapshot();
});
