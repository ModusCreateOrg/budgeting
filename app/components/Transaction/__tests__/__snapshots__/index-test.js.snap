// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="transaction"
>
  <div
    id={4}
  />
</div>
`;
